### ISSUE TITLE
[expo-modules] New architecture incompatibility with expo-drag-drop-content-view library

### ISSUE THREAD

**Issue by AlirezaHadjar** [expo/expo#28040]
### Minimal reproducible example

https://github.com/AlirezaHadjar/ExpoDemo

### What platform(s) does this occur on?

iOS

### Did you reproduce this issue in a development build?

Yes

### Summary

As suggested in [Brent Tweet](https://x.com/notbrent/status/1774931733194088465?s=46&t=HIMTmQZ1qNgFDe3f7YV9kQ) I tried to check if my library [expo-drag-drop-content-view](https://github.com/AlirezaHadjar/expo-drag-drop-content-view) works with the new architecture. Everything works as expected on Android but when I build the app on iOS, the drag and drop feature doesn't work anymore.

If I set these booleans to false everything works as expected again.
app.json:
```json5
"plugins": [
      [
        "expo-build-properties",
        {
          "ios": {
            "newArchEnabled": false
          },
          "android": {
            "newArchEnabled": true
          }
        }
      ]
```

Podfile.properties.json:
```json5
 "newArchEnabled": "false"
 ```

In the first video, the `new arch` is enabled thus drag and drop doesn't work, but in the second one, `new arch` is off and drag and drop works as expected.

https://github.com/expo/expo/assets/57192409/2fbc7b93-f313-4a75-ac6e-c52a4c7cc0ad

https://github.com/expo/expo/assets/57192409/57af5df5-4335-4f4d-98a0-dec204e40d71

I'm wondering if there are any extra steps to follow. Any guide is appreciated 🙂

cc @brentvatne 

### Environment

   System:
      OS: macOS 14.4.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 18.18.0 - ~/.nvm/versions/node/v18.18.0/bin/node
      Yarn: 1.22.21 - ~/.nvm/versions/node/v18.18.0/bin/yarn
      npm: 9.8.1 - ~/.nvm/versions/node/v18.18.0/bin/npm
      Watchman: 2024.03.18.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.15.2 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.4, iOS 17.4, macOS 14.4, tvOS 17.4, visionOS 1.1, watchOS 10.4
      Android SDK:
        API Levels: 27, 28, 29, 30, 31, 33, 33, 34
        Build Tools: 30.0.3, 33.0.0, 33.0.1, 33.0.2, 34.0.0
        System Images: android-28 | Google ARM64-V8a Play ARM 64 v8a, android-33 | Google APIs ARM 64 v8a, android-34 | Google Play ARM 64 v8a
    IDEs:
      Android Studio: 2023.2 AI-232.10227.8.2321.11479570
      Xcode: 15.3/15E204a - /usr/bin/xcodebuild
    npmPackages:
      expo: ^50.0.14 => 50.0.14 
      react: 18.2.0 => 18.2.0 
      react-dom: 18.2.0 => 18.2.0 
      react-native: 0.73.6 => 0.73.6 
      react-native-web: ~0.19.6 => 0.19.10 
    npmGlobalPackages:
      eas-cli: 7.6.2
    Expo Workflow: bare

**Comment by AlirezaHadjar** [comment]
Hi @tsapeta thank you so much for testing this. Yeah, it crashes on iPads with iOS 17 (simulator only). However, it works fine with iOS 16 or the physical device.


I'll try to follow what you said. Hope it works 🤞

**Comment by AlirezaHadjar** [comment]
Just tried out the `0.74 canary` by following your instructions. Unfortunately, It doesn't work. 
You can try it out with these simulators:
![Screenshot 2024-04-18 at 10 07 35 AM](https://github.com/expo/expo/assets/57192409/d9a8d517-b4ea-48b4-b70c-d03ff91beeb4)

**Comment by tsapeta** [comment]
Cool, I just downloaded the SDK for iOS 16.4 and it's no longer crashing, thanks!

I checked it very quickly, but the interesting thing is that it works fine after removing this view https://github.com/AlirezaHadjar/ExpoDemo/blob/master/IDragDropContentView.tsx#L129-L136 (i.e. replacing it with `null`) 😅 I'm guessing that this view somehow captures the drop event 🤔

**Comment by AlirezaHadjar** [comment]
Oh...
I know where the problem is. If you set `pointerEvents='none'` the problem is solved. The problem is that when `<DragDropContentView />` has children components, they block the drag/drop interaction. So to address this, I override the `addSubview` method and the drop/drop interaction to each child as well.
https://github.com/AlirezaHadjar/expo-drag-drop-content-view/blob/af539c8ad2997066efeae4fcfcef7a0f05a11e27/packages/expo-drag-drop-content-view/ios/ExpoDragDropContentView.swift#L25C3-L39C6
But when the `newArch=true` the `addSubview` method is not called anymore for the children components 🤔

**Comment by tsapeta** [comment]
Yeah, looks like turning off pointer events is the easiest way to solve this now. I think I know how to fix it properly in the modules core, but it's a bit more complex problem and I may need to change something in the upstream to make it easier to fix.

Briefly, the `addSubview` is not called because we have to insert children views as siblings. From the layout perspective it's fine, Yoga handles that well and Fabric itself can actually do the same under certain conditions (view flattening). I'm not closing this issue, but it's not something I can address before the SDK 51 release 😢

**Comment by AlirezaHadjar** [comment]
That's totally fine. We can wait for the SDK 51 release. We should find a way to attach drag/drop interactions to children. That was possible with `addSubview` but we should find another way to do so.

Thanks again for putting effort into checking this issue 🙌

**Comment by AlirezaHadjar** [comment]
Hey @tsapeta I tested the new arch today and thanks to this [PR](https://github.com/expo/expo/pull/28829) the issue of subviews is gone 💯 
Although, I'm having another issue caused by this [line](https://github.com/AlirezaHadjar/expo-drag-drop-content-view/blob/17d362f8a80f74587d35d7ed81b9358434a8b5ba/packages/expo-drag-drop-content-view/ios/ExpoDragDropContentView.swift#L25C1-L27C6):

```swift
    override func layoutSubviews() {
        dragDropContentView.frame = bounds
    }
```

This causes touchable children not to work. For example, If I have a `TouchableOpacity` inside my `CustomView`, the `onPress` callback never fires.
Do you have any idea why this is happening?

**Comment by AlirezaHadjar** [comment]
@tsapeta Since there have been many improvements made to the Expo Module and Fabric, is there any solution you have in mind for this issue? I tried Expo sdk 52 with latest expo modules but still facing the issue

**Comment by AlirezaHadjar** [comment]
@Kudo @lukmccall Sorry for pinging you, but I noticed that Tomasz removed himself from the code mention of the Expo module. Would you mind helping me tackle this issue? I’d really appreciate your support!

**Comment by lukmccall** [comment]
Hi @AlirezaHadjar, I'm sorry for the late replay - I was on a Christmas break. I'll take a closer look today/tomorrow.

**Comment by AlirezaHadjar** [comment]
Ty, let me know if you need any more info.

**Comment by lukmccall** [comment]
I’ve identified the source of the problem. There is a difference in how React Native manages children between the two architectures. In the old architecture, the view hierarchy appears as follows:

![image](https://github.com/user-attachments/assets/637093d1-a2a7-448c-b56e-0bb596d7d4f8)

As you can see, the RCTView is positioned above the `DragDropContentView`. However, in the new architecture, the order of these views is reversed, which causes events to be swallowed by the `DragDropContentView`.

**Comment by lukmccall** [comment]
I end up with something like this:
```swift
  override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
    // The event type is incomplete. The number 9 indicates a drag event.
    // If the drag event has been received, we can direct it to the drag-drop content view.
    if event?.type.rawValue == 9 {
      return dragDropContentView.hitTest(point, with: event)
    }
    
    // We know that we're not dealing with a drag event.
    // Other events we want to pass to child views, omitting the `dragDropContentView`.
    // We don't need to use `reactZIndexSortedSubviews` on Fabric, but it seems to be necessary on Paper.
    for subview in reactZIndexSortedSubviews() where subview != dragDropContentView {
      let hit = subview.hitTest(point, with: event)
      if hit != nil {
        return hit
      }
    }
    
    // It doesn't seem to be needed, but I'm not 100% sure.
    return super.hitTest(point, with: event)
  }
```
You need to add this method to the `ExpoDragDropContentView`, as it seems to fix the problem with Fabric. I haven't tested it on Paper yet. It might be better to change the order of the views, but that's not straightforward because React Native uses the index to identify the views later. Changing the order could cause many issues with the shadow tree. 

You can experiment by overriding the `insertSubview(_ view: UIView, at index: Int)` method. There is also the `insertReactSubview(_ subview: UIView!, at atIndex: Int)` method, which should be used with Paper. However, I wasn't able to make my override work as intended.

**Comment by AlirezaHadjar** [comment]
@lukmccall thank you for the detailed answer. It worked like a charm 🎉
[Here](https://github.com/AlirezaHadjar/expo-drag-drop-content-view/blob/main/packages/expo-drag-drop-content-view/ios/ExpoDragDropContentView.swift) is the updated version that works on both Paper and Farbric
